File: helix/thread.py
"""Concurrency lint: shared global writes without mutex protection.

Missing pthread knowledge is a finding, not a fake TSan proof.
If no thread API appears anywhere, return [].
"""

from __future__ import annotations

from collections import defaultdict
from pathlib import Path
import re

from helix import laws
from helix.cparse import strip_comments_keep_lines
from helix.models import Finding, FunctionInfo

_THREAD_API = re.compile(
    r"\b(?:pthread_create|std::jthread|std::thread|CreateThread|thrd_create)\b"
)
_MUTEX = re.compile(r"\b(?:mtx_lock|mtx_timedlock|pthread_mutex)\b")
_GLOBAL_DECL = re.compile(
    r"^(?:\s*(?:static|extern|const|volatile|unsigned|signed|short|long)\s+)*"
    r"(?:(?:struct|enum|union)\s+\w+\s+)?"
    r"(?:\w+\s+)+(?P<name>[A-Za-z_]\w*)\s*(?:=\s*[^;]+)?;",
    re.M,
)


def _file_globals(text: str) -> set[str]:
    depth = 0
    globals_: set[str] = set()
    for line in text.splitlines():
        stripped = line.strip()
        if not stripped or stripped.startswith("#"):
            depth += stripped.count("{") - stripped.count("}")
            continue
        if depth == 0:
            m = _GLOBAL_DECL.match(line)
            if m:
                globals_.add(m.group("name"))
        depth += stripped.count("{") - stripped.count("}")
    return globals_


def _writes_global(body: str, name: str) -> bool:
    pat = re.compile(rf"\b{re.escape(name)}\s*(?:[\[\(]|(?:[+\-*/%&|^]?=))")
    return bool(pat.search(body))


def _has_mutex(body: str) -> bool:
    return bool(_MUTEX.search(body))


def _read_tu_text(functions: list[FunctionInfo]) -> str:
    if not functions:
        return ""
    names = []
    for fn in functions:
        p = Path(fn.file)
        names.extend([
            p,
            Path.cwd() / p,
            Path(__file__).resolve().parents[1] / p,
            Path(__file__).resolve().parents[1] / "testdata" / p.name,
        ])
    for p in names:
        if p.is_file():
            try:
                return strip_comments_keep_lines(
                    p.read_text(encoding="utf-8", errors="replace")
                )
            except OSError:
                continue
    return "\n".join(fn.body for fn in functions)


def _tu_has_thread_api(functions: list[FunctionInfo], file_text: str) -> bool:
    if _THREAD_API.search(file_text):
        return True
    return any(_THREAD_API.search(fn.body) for fn in functions)


def run_thread(functions: list[FunctionInfo]) -> list[Finding]:
    if not functions:
        return []

    by_file: dict[str, list[FunctionInfo]] = defaultdict(list)
    for fn in functions:
        by_file[fn.file].append(fn)

    out: list[Finding] = []
    for rel, fns in by_file.items():
        file_text = _read_tu_text(fns)
        if not _tu_has_thread_api(fns, file_text):
            continue

        globals_ = _file_globals(file_text)
        if not globals_:
            continue

        writers: dict[str, list[FunctionInfo]] = defaultdict(list)
        for fn in fns:
            for g in globals_:
                if _writes_global(fn.body, g):
                    writers[g].append(fn)

        for g, who in writers.items():
            unsync = [fn for fn in who if not _has_mutex(fn.body)]
            if len(unsync) < 2:
                continue
            anchor = unsync[0]
            out.append(
                Finding(
                    stage="thread",
                    status=laws.FAILED,
                    file=rel,
                    function=anchor.name,
                    line=anchor.line,
                    cls="RACE-SHARED",
                    message=(
                        f"global '{g}' written from {len(unsync)} functions "
                        "without mtx_lock/pthread_mutex"
                    ),
                    strength=laws.STRENGTH_FINDS,
                    evidence=g,
                    extra={"global": g, "writers": [fn.name for fn in unsync]},
                )
            )
    return out

File: src/prism/stages/thread.cpp
// Stage thread: file globals written in files that start threads without a mutex.
#include "common.hpp"

namespace prism {
namespace fs = std::filesystem;
using namespace stages_detail;

namespace {
std::set<std::string> file_globals(const std::string& text) {
    int depth = 0;
    std::set<std::string> g;
    static Regex decl(
        "^(?:\\s*(?:static|extern|const|volatile|unsigned|signed|short|long)\\s+)*"
        "(?:(?:struct|enum|union)\\s+\\w+\\s+)?"
        "(?:\\w+\\s+)+(?P<name>[A-Za-z_]\\w*)\\s*(?:=\\s*[^;]+)?;");
    std::istringstream ss(text);
    std::string line;
    while (std::getline(ss, line)) {
        auto stripped = strip(line);
        if (stripped.empty() || stripped.starts_with("#")) {
            depth += static_cast<int>(std::count(stripped.begin(), stripped.end(), '{') -
                                      std::count(stripped.begin(), stripped.end(), '}'));
            continue;
        }
        if (depth == 0) {
            if (auto m = match_at(decl, line)) g.insert(m->named("name"));
        }
        depth += static_cast<int>(std::count(stripped.begin(), stripped.end(), '{') -
                                  std::count(stripped.begin(), stripped.end(), '}'));
    }
    return g;
}

bool writes_global(const std::string& body, const std::string& name) {
    return re_search("\\b" + name + "\\s*(?:[\\[\\(]|(?:[+\\-*/%&|^]?=))", body);
}

}  // namespace

std::vector<Finding> run_thread(const std::vector<FunctionInfo>& functions) {
    if (functions.empty()) return {};
    std::map<std::string, std::vector<FunctionInfo>> by_file;
    for (auto& fn : functions) by_file[fn.file].push_back(fn);
    std::vector<Finding> out;
    static Regex thread_api(
        "\\b(?:pthread_create|std::jthread|std::thread|CreateThread|thrd_create)\\b");
    static Regex mutex_re("\\b(?:mtx_lock|mtx_timedlock|pthread_mutex)\\b");
    for (auto& [rel, fns] : by_file) {
        std::string file_text;
        auto p = locate_source(fns.front());
        if (p) file_text = strip_comments_keep_lines(read_text_file(*p));
        else {
            for (auto& fn : fns) {
                if (!file_text.empty()) file_text += '\n';
                file_text += fn.body;
            }
        }
        bool has_api = thread_api.search(file_text);
        if (!has_api)
            for (auto& fn : fns)
                if (thread_api.search(fn.body)) has_api = true;
        if (!has_api) continue;
        auto globals = file_globals(file_text);
        if (globals.empty()) continue;
        std::map<std::string, std::vector<FunctionInfo>> writers;
        for (auto& fn : fns)
            for (auto& g : globals)
                if (writes_global(fn.body, g)) writers[g].push_back(fn);
        for (auto& [g, who] : writers) {
            std::vector<FunctionInfo> unsync;
            for (auto& fn : who)
                if (!mutex_re.search(fn.body)) unsync.push_back(fn);
            if (unsync.size() < 2) continue;
            auto& anchor = unsync[0];
            auto f = make_find("thread", laws::FAILED, anchor, "RACE-SHARED",
                               "global '" + g + "' written from " + std::to_string(unsync.size()) +
                                   " functions without mtx_lock/pthread_mutex",
                               laws::STRENGTH_FINDS);
            f.file = rel;
            f.evidence = g;
            f.extra["global"] = g;
            nlohmann::json w = nlohmann::json::array();
            for (auto& fn : unsync) w.push_back(fn.name);
            f.extra["writers"] = w.dump();
            out.push_back(std::move(f));
        }
    }
    return out;
}

}  // namespace prism
